sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala

```diff
@@ -4603,6 +4603,15 @@ object SQLConf {
       .booleanConf
       .createWithDefault(true)
 
+  val OPTIMIZE_BETWEEN =
+    buildConf("spark.sql.optimizeBetween")
```

Comment:
It's not only an optimization, but also a correctness fix if the BETWEEN input is nondeterministic. I think a legacy config is better here. How about `spark.sql.legacy.duplicateBetweenInput`? By default it's false, and people can set it to true to restore to the legacy behavior.